WebDriverAgentLib/Categories/XCUIElement+FBUtilities.m

```diff
@@ -147,6 +110,76 @@ - (nullable XCElementSnapshot *)fb_snapshotWithAttributes {
   return snapshotWithAttributes;
 }
 
++ (BOOL) fb_newSnapshotAPIIsSupported
+{
+  static dispatch_once_t newSnapshotIsSupported;
+  static BOOL result;
+  dispatch_once(&newSnapshotIsSupported, ^{
+    result = [(NSObject *)[FBXCTestDaemonsProxy testRunnerProxy] respondsToSelector:@selector(_XCT_requestSnapshotForElement:attributes:parameters:reply:)];
+  });
+  return result;
+}
+
+- (void) fb_requestSnapshot:(XCAccessibilityElement *)accessibilityElement reply:(void (^)(XCElementSnapshot *, NSError *))block
+{
+  static NSDictionary *defaultParameters;
+  static dispatch_once_t initializeAttributesAndParametersToken;
+  static NSArray *axAttributes;
+  BOOL useNewSnapshotAPI = [XCUIElement fb_newSnapshotAPIIsSupported];
+  dispatch_once(&initializeAttributesAndParametersToken, ^{
+    defaultParameters = [FBXCAXClientProxy.sharedClient defaultParameters];
+    axAttributes = [self fb_createAXAttributes:!useNewSnapshotAPI];
+  });
+  id<XCTestManager_ManagerInterface> proxy = [FBXCTestDaemonsProxy testRunnerProxy];
+  if (useNewSnapshotAPI) {
+    [proxy _XCT_requestSnapshotForElement:self.lastSnapshot.accessibilityElement
+                               attributes:axAttributes
+                               parameters:defaultParameters
+                                    reply:block];
+  } else {
+    [proxy _XCT_snapshotForElement:self.lastSnapshot.accessibilityElement
+                        attributes:axAttributes
+                        parameters:defaultParameters
+                             reply:block];
+  }
+}
+
+- (NSArray *) fb_createAXAttributes: (BOOL)asNumber
+{
+  // Names of the properties to load. There won't be lazy loading for missing properties,
+  // thus missing properties will lead to wrong results
+  NSArray<NSString *> *propertyNames = @[
```

Comment:
is probably makes sense to make this array static and fill it with values inside dipatch_once block

Reply:
No need to do extra optimisation here.  `fb_createAXAttributes` will be called once anyways by fb_requestSnapshot. I don't like using dispatch_once a lot. It is a bad practice

Reply:
with the current approach we create a new array object every time we enter this method (e.g take a snapshot). Putting it into static will save some milliseconds on each call

Reply:
I can optimise that later on but this is seriously not a big issue